### PR TITLE
Bug 1878972: adm release mirror: add architecture information

### PR DESCRIPTION
**Pull request by jottofar**
Retrieve architecture from image release information and use in naming.

### CODE REVIEW

File: pkg/cli/admin/release/info.go
Comment (wking) [inline pkg/cli/admin/release/info.go:503]: What is the benefit to storing this here, after we'e already successfully called `o.describeImage(release)`?

File: pkg/cli/admin/release/mirror.go
Comment (wking) [inline pkg/cli/admin/release/mirror.go:64]: Image architectures are [standardized][1] as Go arch strings, so maybe:

```go
// archMap maps Go architecture strings to OpenShift alternatives, when OpenShift prefers a different identifier than Go.
```

Or some such?

[1]: https://github.com/opencontainers/image-spec/blame/v1.0.1/config.md#L103-L106

File: pkg/cli/admin/release/info.go
Comment (wking) [reply pkg/cli/admin/release/info.go:503]: Ah, for the `infoOpts.Run` call in `mirror.go`, when we are loading from a pullspec.  I'm... not thrilled about that, but it's an existing approach, so we should not block this PR on cleaning up that API.

File: pkg/cli/admin/release/mirror.go
Comment (wking) [inline pkg/cli/admin/release/mirror.go:477]: I think we can drop this hunk and the `info.go` changes.  We should be able to grab the arch in `ImageMetadataCallback` below from [`config.Architecture`][1], with something like:

```go
extractOpts.ImageMetadataCallback = func(m *extract.Mapping, dgst, contentDigest digest.Digest, config *dockerv1client.DockerImageConfig) {
  releaseDigest = contentDigest.String()
  verifier.Verify(dgst, contentDigest)
  if config != nil && config.Architecture != "" {
    if val, ok := archMap[config.Architecture]; ok {
      archExt = "-" + val
    } else {
      archExt = "=" + config.Architecture
    }
  }
}
```

[1]: https://github.com/openshift/oc/blob/0ccbf0633fd374af88f444d9e5cc6bd595933670/vendor/github.com/openshift/library-go/pkg/image/dockerv1client/types.go#L91

File: pkg/cli/admin/release/info.go
Comment (wking) [reply pkg/cli/admin/release/info.go:503]: [This][1] might be one way to avoid it.

[1]: https://github.com/openshift/oc/pull/646#discussion_r527400619

**Review by wking — APPROVED** [review]
After building `oc` from e575833cd:

```console
$ ./oc adm release mirror --to-dir=mirror quay.io/openshift-release-dev/ocp-release:4.6.4-x86_64
...
    manifests:
      sha256:003d999d37f8d2c043ad9370eaf1e7533b9306bcebea47d62546b6456b0f6924 -> 4.6.4-x86_64-grafana
      sha256:00900d48c5796ecb8c0599ab6a0946347947dbcd2acc883665240c2ec9b33fd5 -> 4.6.4-x86_64-pod
...
```

I didn't have enough local disk space for that to complete so I could check `mirror/v2/openshift/release/manifests/`, but those should now include the arch suffix too.

/lgtm
/bugzilla cc-qa
/hold

(Trying out the pre-merge testing workflow, we can pull the hold once QE verifies the change).

**Comment by jottofar** [comment]
Requesting review from QA contact:
/cc @jianlinliu

**Comment by jianlinliu** [comment]
lgtm. Once merged, will run a real testing for it.

**Comment by stbenjam** [comment]
This ended up breaking baremetal IPI, we manually built a pullspec for a mirrored payload image since it wasn't easily extractable and adding the arch broke our assumption. Could we add the metal job here? We use `oc` *a lot* and changes like this have broken us a few times.

I doubt customers are hitting this one... but it would've been nice to get a heads up this was a breaking change for us.

See: https://github.com/openshift/release/pull/14244

**Comment by jianlinliu** [comment]
@stbenjam if this is breaking your testing, can you add comments in 1878972 bug, and assign it back.

**Comment by stbenjam** [comment]
We've fixed the issue in our CI. In this case we were doing something kind of weird. But we use `oc` a lot so having a heads up on this would've made it easier.

**Pull request merged**
